### PR TITLE
Make zsock_select() to support microsecond resolution

### PR DESCRIPTION
**Pull request by jukkar**
This fixes #36072

### CODE REVIEW

File: subsys/net/lib/sockets/sockets_select.c [hunk 1]
```diff
@@ -197,3 +200,75 @@ int zsock_select(int nfds, zsock_fd_set *readfds, zsock_fd_set *writefds,
...
+			       zsock_fd_set *writefds,
+			       zsock_fd_set *exceptfds,
+			       struct zsock_timeval *timeout)
+{
+	zsock_fd_set *readfds_copy, *writefds_copy, *exceptfds_copy;
+	size_t fds_size;
+	int ret;
+
+	/* Copy fds array from user mode */
+	if (size_mul_overflow(nfds, sizeof(zsock_fd_set), &fds_size)) {
```
Comment (rlubos) [inline subsys/net/lib/sockets/sockets_select.c:218]: I don't think I understand the purpose of this multiplication. Unless I miss something, the `readfds`/`writefds`/`exceptfds` should always point to a single instance of `zsock_fd_set` structure, regardless of the `nfds` parameter. `nfds` just speicifes the range of the fds to verify in the `zsock_fd_set` bitfield.

File: subsys/net/lib/sockets/sockets_select.c [hunk 2]
```diff
@@ -197,3 +200,75 @@ int zsock_select(int nfds, zsock_fd_set *readfds, zsock_fd_set *writefds,
 
 	return num_selects;
 }
+
+#ifdef CONFIG_USERSPACE
+static int z_vrfy_zsock_select(int nfds, zsock_fd_set *readfds,
+			       zsock_fd_set *writefds,
+			       zsock_fd_set *exceptfds,
+			       struct zsock_timeval *timeout)
```
Comment (rlubos) [inline subsys/net/lib/sockets/sockets_select.c:211]: Shouldn't this function copy the `timeout` parameter as well?

File: subsys/net/lib/sockets/sockets_select.c (hunk 1 above)
Comment (jukkar) [reply subsys/net/lib/sockets/sockets_select.c:218]: I copied this from poll implementation, but that was not a proper thing to do. I will change this.

File: subsys/net/lib/sockets/sockets_select.c (hunk 2 above)
Comment (jukkar) [reply subsys/net/lib/sockets/sockets_select.c:211]: > Shouldn't this function copy the `timeout` parameter as well?

Yep, it should, will fix.

**Review by rlubos — APPROVED** [review]

**Comment by jukkar** [comment]
ping @pfalcon do you have time for a review?

**Review by pfalcon — COMMENTED** [review]
select() is a "legacy" POSIX function with various drawbacks. It was never intended to be part of "core" POSIX functionality in Zephyr, and was added solely to help to port existing 3rd-party code to Zephyr. It explicitly was implemented "in userspace", stacked on top of poll() implementation to not grow surface size of Zephyr syscall API and associated issues (code size, testability, security). And not the last was also the idea to discourage its usage in the code Zephyr natively written from Zephyr (any curious user who would check its implementation would see that it's implemented in terms of poll() and would use poll() directly).

I don't find arguments in linked https://github.com/zephyrproject-rtos/zephyr/discussions/36072 convincing, actually there're no arguments, just statement of user's preference to select() based on theoretical better timing resolution. It's unclear whether that theoretical difference leads to any tangible benefits - after all there's a reason why it was considered enough to have millisecond-resolution for poll() in general for POSIX systems, and unlikely any benefit in case of Zephyr in particular.

poll() is otherwise better match for Zephyr's architecture, it matches Zephyr's native z_poll() API. But even then it requires O(n) conversation for arguments/results (select requires the same, but with higher constant multiplier).

The only way forward to improve efficiency is to implement epoll() Linux-inspired POSIX extension call, backed by compatible native k_epoll() implementation to avoid constant O(n) data structure marshalling back and forth. Note that Linux epoll() takes timeout in milliseconds, because again, it apparently was found to impractical to provide higher resolution for I/O scheduling (aka: 1 microsecond passed to select() will likely be rounded up to much higher delay anyway).

Based on the above (this patch violating original requirement of Zephyr POSIX subsys core being small and avoid duplicate functionality), I'd abstain from giving it +1. Otherwise, assuming it needs to be done, the implementation here is reasonable, so I'm not -1 on it either.

**Comment by jukkar** [comment]
I partly agree that poll() is a better API to use. This PR is just fixing the ms vs us "confusion" with the timeout value.

**Comment by pfalcon** [comment]
> I partly agree that poll() is a better API to use. This PR is just fixing the ms vs us "confusion" with the timeout value.

... and grows number of kernel syscalls and associated argchecking (== security related) code. I agree that's unavoidable direction in general, but in this particular case, what I wanted to avoid with the original impl. So I hope you forgive my not jumping with +1 it, I'm sure someone else will approve it w/o issues ;-).

**Review by pfalcon — APPROVED** [review]
Ok, I'm not sure if it's vacations starting or something, but as nobody else approved this, let me do it now - again it's a reasonable way to do that, for as long as it needs to be done at all.

**Pull request merged**